### PR TITLE
Hermite asymptotic

### PR DESCRIPTION
**Pull request by raoulbq**
Implementation of an asymptotic algorithm to compute Gauss-Hermite quadrature points and weights of very large order n. The algorithm has linear time complexity and is optimal. Numerically it starts paying off around n = 200 and can go up to at tens of thousands of points (I tried up to 50'000). The current Golub-Welsch code can work reliably up to about 210 and we hand over at 200. This probably supersedes PR #3987.

### CODE REVIEW

**Comment by raoulbq** [comment]
Some plots showing the runtimes of old ("GW") and new ("ASY") code.
scaling_asy
scaling_asy_log
scaling_asy_small

**Comment by raoulbq** [comment]
Numerical errors (absolute and relative) of the weights and nodes. The reference values were computed by arbitrary precision arithmetic including rigorous errors bounds.
error_weights_rel
error_weights_abs
error_nodes_rel
error_nodes_abs

**Comment by charris** [comment]
I'm fascinated that there is a use case for Gauss-Hermite integration of such high order. What are you doing that needs it?

**Comment by argriffing** [comment]
This seems pretty great.  At the risk of commenting on the bike shed instead of on the plans for the nuclear power plant, could this use some unit tests that specifically target the asymptotic and Golub-Welsch code paths?  It looks like N is tested for {5, 25, 100} in https://github.com/scipy/scipy/blob/master/scipy/special/tests/test_orthogonal.py.  Maybe tests with 'slow' annotations could be added for N above the crossover point, or explicit tests for each path?  This also looks like a good candidate for `asv` benchmarks, but I guess that could fall outside of the scope of the PR.

**Comment by charris** [comment]
PEP8 failures. If you click through to the failing test you can find a list of violations.

**Comment by argriffing** [comment]
I branched the code to fix the PEP8 failures, but now I see that they are mostly false positives in my opinion.  I am happy to ignore this red X or to loosen the PEP8 bot settings.

**Comment by raoulbq** [comment]
I tried adding one more term in the series expansion compared to the original paper.
At the expense of being a little bit slower, we get accurate results already for n=100.

**Comment by raoulbq** [comment]
Runtime is about +0.1s for 50'000 nodes.
scaling_asy
scaling_asy_log
scaling_asy_small

**Comment by raoulbq** [comment]
Numerical errors:
error_weights_rel
error_weights_abs
error_nodes_rel
error_nodes_abs

**Comment by raoulbq** [comment]
While the nodes show extremely fast convergence, the weights have some issues.
Anyway, I'm not sure if this additional term is worth the effort.

**Comment by raoulbq** [comment]
> Maybe tests with 'slow' annotations could be added for N above the crossover point, or explicit tests for each path?  

Even for large N it is not slow any more.

I tried to add explicit tests. I think they fail because
the evaluation of the Hermite polynomials goes wrong.

> This also looks like a good candidate for `asv` benchmarks, but I guess that could fall outside of the scope of the PR.

I don't know about this.

**Comment by raoulbq** [comment]
> I branched the code to fix the PEP8 failures, but now I see that they are mostly false positives in my opinion.  I am happy to ignore this red X or to loosen the PEP8 bot settings.

I know the code is not PEP8 compliant. While PEP8 provides good default rules,
they fail for math code. There it might make more sense to format equations
differently. I can reformat the code if necessary.

**Comment by raoulbq** [comment]
> I'm fascinated that there is a use case for Gauss-Hermite integration of such high order. What are you doing that needs it?

I use Hermite functions as basis for solving PDEs
and I'd like to go up to about N=1000. Evaluation
of inner products then might need higher order rules.
The relatively low bound of 200 posed strong restrictions.

There might be no use case for N > 1000 but since
the algorithm can do it, we get it for free.

**Comment by ewmoore** [comment]
For those tests where evaluating the hermite polynomials fail, I'd calculate the appropriate roots via some other method, and include them in the file or as a npz file that you can use.  You shouldn't need to make any changes to `verify_gauss_quad` since you can define a dummy function that just returns your roots.

Additionally, Any function that is user facing should be prefixed with a `_`.

File: scipy/special/orthogonal.py [hunk 1]
```diff
@@ -54,6 +54,8 @@
   la_roots               -- Generalized Laguerre
   h_roots                -- Hermite
   he_roots               -- Hermite (unit-variance)
+  h_roots_asy            -- Hermite, asymptotic version
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Is there a good reason to expose these?  Shouldn't this all just go through `h_roots`?  The method used to produce the roots shouldn't really matter to someone using the function and you changes below already switch methods at a reasonable point.

File: scipy/special/orthogonal.py [hunk 2]
```diff
@@ -498,6 +503,8 @@ def h_roots(n, mu=False):
     :math:`H_n(x)`.  These sample points and weights correctly integrate
     polynomials of degree :math:`2*n - 1` or less over the interval
     :math:`[-inf, inf]` with weight function :math:`f(x) = e^{-x^2}`.
+    For n larger than 200 an optimal asymptotic algorithm is used which
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Probably in a Notes section instead.

File: scipy/special/orthogonal.py [hunk 3]
```diff
@@ -525,12 +532,277 @@ def h_roots(n, mu=False):
...
+    else:
+        nodes, weights = h_roots_asy(n)
+        if mu:
+            return nodes, weights, sum(weights)
+        else:
+            return nodes, weights
+
+
+def compute_tauk(n, k, steps=5):
+    """Helper function for lemma 3.1
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: This comment doesn't tell me anything. I don't know that every helper function needs a full docstring in the usual style, but it ought to have something at least moderately helpful.

File: scipy/special/orthogonal.py [hunk 4]
```diff
@@ -525,12 +532,277 @@ def h_roots(n, mu=False):
...
+    # Combine
+    iv = sqrt(hstack([xasq, xbsq]))
+    # Central node is always zero
+    if n % 2 == 1:
+        iv = hstack([0.0, iv])
+    return iv
+
+
+def pbcf(n, theta):
+    """Asymptotic series expansion of parabolic cylinder function.
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Does our parabolic cylinder function have this expansion?  I haven't looked or tried, but do you think it would be worth adding?  Could be a separate PR.

File: scipy/special/orthogonal.py (hunk 1 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: Probably not except of making things explicit. Should I remove the export?

File: scipy/special/orthogonal.py (hunk 4 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: I did not look at the included functions. The notation in scipy[1] seems to differ from standard references like [2,3]. Is it correct that scipy has no `U(a, z)`?

[1]: http://docs.scipy.org/doc/scipy/reference/special.html#parabolic-cylinder-functions
[2]: http://dlmf.nist.gov/12
[3]: http://dlmf.nist.gov/12.1

File: scipy/special/orthogonal.py (hunk 3 above)
Comment (argriffing) [inline scipy/special/orthogonal.py]: I like docstrings that give this information, I guess it's for Townsend, A. and Trogdon, T. and Olver, S. (2014) whose reference is defined in the docstring for a different function.  Apparently it's not possible to cite references like `[1]_` across docstrings.

Comment (raoulbq) [inline scipy/special/orthogonal.py]: > I guess it's for Townsend, A. and Trogdon, T. and Olver, S. (2014) whose reference is defined in the docstring for a different function.

Of course it is for this paper that defines the whole algorithm.
Where should this top-level reference go? Up to the other two
in this module?

I'll write better docstrings tomorrow. I agree this is to little to be useful.

File: scipy/special/orthogonal.py (hunk 4 above)
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Scipy has `D_v(z)` as `special.pbdv` which is a different notation for `U`.  See 12.2.5 from your dlmf link.

File: scipy/special/orthogonal.py (hunk 2 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > Probably in a Notes section instead. 

Done

File: scipy/special/orthogonal.py (hunk 3 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > This comment doesn't tell me anything.

Should be much better now.

**Comment by raoulbq** [comment]
Except pep8, what is missing here?

File: scipy/special/orthogonal.py [hunk 5]
```diff
@@ -525,12 +540,454 @@ def h_roots(n, mu=False):
...
+    """
+    sm = 3*pi*(4*m-1) / 8.0
+    coeffs = row_stack([1.0, 5.0/48.0, -5.0/36.0, 77125.0/82944.0,
+                        -108056875.0/6967296.0, 162375596875.0/334430208.0])
+    smp = column_stack([ones_like(sm), sm**(-2), sm**(-4), sm**(-6), sm**(-8), sm**(-10)])
+    am = -sm**(2.0/3.0) * dot(smp, coeffs).reshape((-1,))
+    return am
+
+
+def airy_root(m):
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Is this function necessary?  Will `special.ai_zero` work here? The existing function and your new function seem to give me pretty close results (~1e-12) upto at least `n = 25000`.  If the existing function is not accurate enough, that would also be good to know.

Comment (pv) [inline scipy/special/orthogonal.py]: `special.ai_zero` is about 10x slower than this, presumably because it does not just apply the asymptotic expansion.
However, it would probably be fairly easy to speed that up.

File: scipy/special/orthogonal.py [hunk 6]
```diff
@@ -525,12 +540,454 @@ def h_roots(n, mu=False):
...
+             -4.0879494441309706166,
+             -5.5205598280955510591,
+             -6.7867080900717589988,
+             -7.9441335871208531231,
+             -9.0226508533409803802,
+             -10.040174341558085931,
+             -11.008524303733262893,
+             -11.936015563236262517,
+             -12.828776752865757200])
+    # Note: This specialized code is much faster than 'ai_zeros'
```
Comment (raoulbq) [inline scipy/special/orthogonal.py]: Yes, this function is necessary because it is much faster.
You can try what happens if we take the scipy routine.
I looked into the FORTRAN code and it seems to apply the same asymptotic formula but then doing some more stuff
in a second loop.

Comment (pv) [inline scipy/special/orthogonal.py]: The problem in the fortran code seems to be that the airy function evaluation does too much work in the asymptotic region. Eg. it computes expansion coefficients up to order 40 even if they are not needed, and uses too many of them in the far asymptotic region.

Comment (raoulbq) [inline scipy/special/orthogonal.py]: Hmm, taking a closer look again, it seems not to matter after all. With all other improvements wrt the first version in place, it is as fast as my own routine.
asy

**Comment by raoulbq** [comment]
The airy zero step computes 4 times the data necessary. Probably it would be a good idea to decouple
computation of zeros, zeros of derivative and function values for airy functions.

**Comment by pv** [comment]
See gh-4621 for airy zeros speedup

File: scipy/special/orthogonal.py [hunk 7]
```diff
@@ -525,12 +541,413 @@ def h_roots(n, mu=False):
...
+    tauk = compute_tauk(n, k)
+    sigk = cos(0.5*tauk)**2
+    a = n % 2 - 0.5
+    nu = 4.0*floor(n/2.0) + 2.0*a + 2.0
+    # Initial approximation of Hermite roots (square)
+    xksq = nu*sigk - 1.0/(3.0*nu) * (5.0/(4.0*(1.0-sigk)**2) - 1.0/(1.0-sigk) - 0.25)
+    return xksq
+
+
+def compute_am(m):
```
Comment (pv) [inline scipy/special/orthogonal.py]: Unused function

**Comment by pv** [comment]
In case the parabolic function asy-expansion is implemented here because the version in scipy.special is lacking, those improvements could be made there also. Can be done after merging this, too, provided a ticket is opened.

**Comment by raoulbq** [comment]
> In case the parabolic function asy-expansion is implemented here

This will be the case, because the function is in specfun.f
and I don't know FORTRAN well enough to write it.

> Can be done after merging this, too, provided a ticket is opened.

Where should that be done? Github issue tracker?

**Comment by pv** [comment]
@raoulbq: yes: gh-4626. It would be useful if you comment there more specifically what doesn't work as expected.

**Comment by raoulbq** [comment]
> It would be useful if you comment there more specifically what doesn't work as expected.

Well, actually I implemented along the formulas in the paper and never tried the routine from scipy.

**Comment by raoulbq** [comment]
What is missing to merge this PR?

File: scipy/special/tests/test_orthogonal.py [hunk 8]
```diff
@@ -397,6 +397,32 @@ def test_h_roots():
...
+            H[1,:] = sqrt(2.0) * nodes * H[0,:]
+            for k in xrange(2, n):
+                H[k,:] = sqrt(2.0/k) * nodes * H[k-1,:] - sqrt((k-1.0)/k) * H[k-2,:]
+        return H
+
+    # This tests only the nodes
+    def test(N, rtol=1e-15, atol=1e-14):
+        x, w = orth.h_roots_asy(N)
+        H = hermite_recursion(N+1, x)
+        assert_allclose(H[-1,:], np.zeros(N), rtol, atol)
```
Comment (ewmoore) [inline scipy/special/tests/test_orthogonal.py]: How about checking that the weights sum up appropriately here?

Comment (ewmoore) [inline scipy/special/tests/test_orthogonal.py]: is the `mu` return from `h_roots` check for the new code path anywhere?

File: scipy/special/orthogonal.py [hunk 9]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+    x = mu * cos(theta)
+    # Central node is always zero
+    if n % 2 == 1:
+        x[0] = 0.0
+    # Compute weights
+    w = exp(-x**2) / (2.0*ud**2)
+    return x, w
+
+
+def h_roots_asy(n):
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: All of the functions that are not exposed need to start with an underscore.

File: scipy/special/orthogonal.py [hunk 10]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+
+    See Also
+    --------
+    initial_nodes
+    h_roots_asy
+    """
+    a = n % 2 - 0.5
+    nu = 4.0*floor(n/2.0) + 2.0*a + 2.0
+    # Airy roots by approximation
+    ak = flipud(specfun.airyzo(k, 1)[0])
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: The docstring claims that `k` is an array (and it looks as though you pass an array in down below), but the first argument to `airyzo` is an int, not an array?  When I call `airyzo` with an array, it simply uses the first element. 
```
In [68]: special.specfun.airyzo(np.arange(3,10), 1)[0]
Out[68]: array([-2.33810741, -4.08794944, -5.52055983])

In [69]: special.specfun.airyzo(3, 1)[0]
Out[69]: array([-2.33810741, -4.08794944, -5.52055983])
```
Is this the need for `flipud` here and below?

File: scipy/special/orthogonal.py [hunk 11]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+    .. [townsend.trogdon.olver-2014]
+       Townsend, A. and Trogdon, T. and Olver, S. (2014)
+       *Fast computation of Gauss quadrature nodes and
+       weights on the whole real line*. ArXiv 1410.5286.
+    """
+    m = int(n)
+    if n < 1 or n != m:
+        raise ValueError("n must be a positive integer.")
+
+    iv = initial_nodes(n)
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: This should probably be `m` instead of `n`.  In `initial_nodes` you are doing `floor(n+0.5)` to get it back to int again.  Additionally, since this is a routine only called by `h_roots` you don't need to check and raise a `ValueError` here, the parameter passed here should already have been validated.

File: scipy/special/orthogonal.py [hunk 12]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+    a3 = 0.12822657455632716049
+    a4 = 0.29184902646414046425
+    a5 = 0.88162726744375765242
+    b0 = 1.0
+    b1 = -0.14583333333333333333
+    b2 = -0.09874131944444444444
+    b3 = -0.14331205391589506173
+    b4 = -0.31722720267841354810
+    b5 = -0.94242914795712024914
+    a = (a0, a1, a2, a3, a4, a5)
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Why make a tuple here?  You could just as well use `a0, a1` etc below.

File: scipy/special/orthogonal.py [hunk 13]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+    u = (u0, u1, u2, u3, u4, u5)
+    v = (v0, v1, v2, v3, v4, v5)
+    # Airy Evaluation (Bi and Bip unused)
+    Ai, Aip, Bi, Bip = airy(mu**(4.0/6.0) * zeta)
+    # Prefactor for U
+    P = 2.0*sqrt(pi) * mu**(1.0/6.0) * phi
+    # Terms for U
+    # http://dlmf.nist.gov/12.10#E42
+    A0 = b[0]*u[0]
+    A1 = (b[2]*u[0] + phi**6*b[1]*u[1] + phi**12*b[0]*u[2]) / zeta**3
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Repeatedly raising `phi` to the same power probably makes a measurable impact on the speed here, you could do as you did above with `ctp`.

File: scipy/special/orthogonal.py [hunk 14]
```diff
@@ -525,12 +540,383 @@ def h_roots(n, mu=False):
...
+        Quadrature nodes
+    weights : ndarray
+        Quadrature weights
+
+    See Also
+    --------
+    h_roots
+
+    References
+    ----------
```
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Also add this reference to `h_roots`.

File: scipy/special/orthogonal.py (hunk 10 above)
Comment (ewmoore) [inline scipy/special/orthogonal.py]: Is this using something that doesn't appear `special.ai_zeros`?  If not, just call the top level function.

**Comment by ewmoore** [comment]
I think this is getting pretty close.  There are places where the same calculation is done repeatedly that I haven't comment on, but could also be changed. I don't feel strongly either way.

Does Scipy have a preference between `flipud(arr)` and `arr[::-1]` when arr is 1D? I admit to preferring the later.

File: scipy/special/orthogonal.py (hunk 9 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > All of the functions that are not exposed need to start with an underscore.

Done.

File: scipy/special/orthogonal.py (hunk 13 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > Repeatedly raising `phi` to the same power probably makes a measurable impact

Thanks for the hint. Yes I can measure the difference, computation time falling
from 0.45 to about 0.38 for 50'000 nodes.

File: scipy/special/orthogonal.py (hunk 10 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > Is this using something that doesn't appear `special.ai_zeros`?

Just bypassing the checks.

>  If not, just call the top level function.

I had some issues with cyclic imports when using the top
level ai_zeros which is defined in `basic.py` and basic
imports `orthogonal.py`. I can not import it from special
it seems.

File: scipy/special/orthogonal.py (hunk 11 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > This should probably be `m` instead of `n`.

Done.

> In `initial_nodes` you are doing `floor(n+0.5)`

There is no `floor(n+0.5)`?

File: scipy/special/orthogonal.py (hunk 10 above)
Comment (raoulbq) [inline scipy/special/orthogonal.py]: > The docstring claims that `k` is an array (and it looks as though you pass an array in down below)

That is true, yes. It is the index of all Hermite nodes that are
computed by this approximation.

> but the first argument to `airyzo` is an int, not an array?  When I call `airyzo` with an array, it simply uses the first element. 

Right. But it will return all Ai zeros up to the scalar index k.
My array k now comes with largest element first hence it works.
Maybe passing k.max() would be more clear.

> Is this the need for `flipud` here and below? 

Not directly.

File: scipy/special/tests/test_orthogonal.py (hunk 8 above)
Comment (raoulbq) [inline scipy/special/tests/test_orthogonal.py]: > How about checking that the weights sum up appropriately here? 

Done.

Comment (raoulbq) [inline scipy/special/tests/test_orthogonal.py]: > is the `mu` return from `h_roots` check for the new code path anywhere?

Done.

**Comment by raoulbq** [comment]
> There are places where the same calculation is done repeatedly that I haven't comment on, but could also be changed.

The only remaining cases are `zeta` but each power occurs only twice.
All other variables are scalars that don't matter.

> Does Scipy have a preference between `flipud(arr)` and `arr[::-1]` when arr is 1D? I admit to preferring the later.

I removed the flip functions in the top-most commit.

**Comment by raoulbq** [comment]
index
indexa
indexb
indexc
indexd

**Comment by raoulbq** [comment]
Any additional review comments?

**Comment by raoulbq** [comment]
Can we merge this or are there still missing bits?

**Comment by rgommers** [comment]
@raoulbq I haven't followed this PR in detail, but do want to say I like the detailed plots - quite easy to figure out what this PR improved. Thanks!

Guess @argriffing or @ewmoore will merge this soon-ish.

**Comment by argriffing** [comment]
> Guess @argriffing or @ewmoore will merge this soon-ish.

I haven't kept up with @pv's airy zeros stuff or with @ewmoore's comments.  If they think it is good to merge then I don't have any objection!

**Comment by ewmoore** [comment]
I've looked through this again and it look okay to me.  It should be squashed down some though.

**Comment by raoulbq** [comment]
@ewmoore Should I combine the commits? By how much?

**Comment by pv** [comment]
IMO, good to merge as is.

**Comment by ewmoore** [comment]
Okay then.  In it goes as is.  Thanks @raoulbq!

**Comment by raoulbq** [comment]
Thank you for merging.
